README.md
[![License](https://img.shields.io/github/license/Matticusau/pr-helper.svg?style=flat-square)](LICENSE)
[![Last commit](https://img.shields.io/github/last-commit/Matticusau/pr-helper.svg?style=flat-square)](https://github.com/heinrichreimer/action-github-changelog-generator/commits)
[![Latest tag](https://img.shields.io/github/tag/Matticusau/pr-helper.svg?style=flat-square)](https://github.com/heinrichreimer/action-github-changelog-generator/releases)
[![Issues](https://img.shields.io/github/issues/Matticusau/pr-helper.svg?style=flat-square)](https://github.com/heinrichreimer/action-github-changelog-generator/issues)
[![Pull requests](https://img.shields.io/github/issues-pr/Matticusau/pr-helper.svg?style=flat-square)](https://github.com/heinrichreimer/action-github-changelog-generator/pulls)

# pr-helper

Extremely powerful [GitHub Action](https://github.com/features/actions) to streamline management of PRs through automation of common tasks. Very versatile with plenty of configuration settings to adapt to many different implementations.

Currently supports the following functionality:

- Welcome message on new PR
- Automatic Label assignment and removal
  - Based on Review criteria (required/provided)
  - If PRs qualify for automatic merge
- PR Comment automation
  - Welcome message with instructions on new PRs
  - Assign a label indicating if the PR is ready or on hold based on key words
- Pull Request Reviewers
  - Automatically assign reviewers from YAML front matter
  - Name matching to github username via Jekyll or DocFX Author/People YAML file. More details [here](./docs/FrontMatter.md).
- Pull Request Merge
  - Automatically merge when criteria is met
  - Respect requested reviews (_i.e. CODEOWNERS_)
  - Respect a minimal review count (_for now configured by input param, repo setting on roadmap_)
  - Delete branch (head ref) on merge

## Events

The Action can respond to the following [workflow events](https://help.github.com/en/actions/reference/events-that-trigger-workflows):

- pull_request
- pull_request_review
- issue_comment
- schedule

## Inputs

Set the following inputs in the workflow file

### `repo-token`

**Required** The token to use for github authentication. Recommend using `${{ secrets.GITHUB_TOKEN }}`. If additional access is required use a PAT/Secret and set it as a secret. More info see [here](https://help.github.com/en/actions/configuring-and-managing-workflows/authenticating-with-the-github_token).

> If you have enabled Branch Protection rules then a PAT/Secret will need to be configured.

### `enable-prmerge-automation`

**Required** Set to true to enable the auto merge capabilities

### `enable-prcomment-automation`

**Required** Set to true to enable the PR/Issue comment automation

### `enable-prlabel-automation`

**Required** Set to true to enable the PR label automation

### `enable-prreviewer-frontmatter`

Set to true to enable reviewers to be set from owner in YAML front matter.

### `enable-welcomemessage`

**Required** Set to true to automatically send a welcome message to new contributors

### `welcome-message`

The custom welcome message to send to new contributors

Requires `enable-welcomemessage: true`

### `prmerge-requireallchecks`

Set to true if all checks need to complete before auto merging

### `prmerge-requirereviewcount`

Should match the setting in your GitHub repo. Set it to -1 to disable.

### `prmerge-method`

The method to use when merging the PRs

### `prmerge-pathcheck`

Set to true to require a path check for auto merge capabilities

Requires `enable-prmerge-automation: true`

### `prmerge-allowpaths`

Provide the path globs which will allow auto merge. JSON object as string, example format {"any":["docs/**"]}.

Requires `prmerge-pathcheck: true`

### `prmerge-deletebranch`

If true then the branch will be deleted on merge

Requires `enable-prmerge-automation: true`

### `prmerge-deletebranch-config`

Provide the branch patterns which will allow/deny auto delete on merge. JSON object as string, example format {"deny":["dev"]}.

Requires `prmerge-deletebranch: true`

### `prreviewer-authorkey`

The key in the YAML front matter to define the article author(s), who will be assigned as reviewers. Defaults to `author`

```yml
---
title: My great article
author: octocat
---
```

Requires `enable-prreviewer-frontmatter: true`

### `prreviewer-githubuserfromauthorfile`

When set to `true` enables the lookup of the author from the Jekyll style Author YAML file

Requires `enable-prreviewer-frontmatter: true`

### `prreviewer-authorfilepath`

Provides the ability to configure the path to the Jekyll authors YAML file to use in lookup. Default is `docs/_data/authors.yml`

Requires `enable-prreviewer-frontmatter: true` and `prreviewer-githubuserfromauthorfile: true`

### `prreviewer-bypassforfileowner`

Provides the ability to by pass review checks when all the changed files are owned by the author of the PR.
If you configure [Branch Protection](https://docs.github.com/en/github/administering-a-repository/about-protected-branches) within GitHub security, then it may block this setting from working. Configuring the `prmerge-requirereviewcount` setting in it's place will not affect this setting.

Requires `enable-prreviewer-frontmatter: true`

### `prlabel-default`

The initial label to set on PRs when first created

### `prlabel-ready`

The label to use when the PR has been signed off and ready for merge

### `prlabel-onhold`

The label to use when the PR is on hold

### `prlabel-reviewrequired`

The label to use when the PR requires reviews

### `prlabel-automerge`

The label to set on PRs when it qualifies for automatic merge by this action

## Outputs

None

## Example usage

Create the following file within your repo as `.github/workflows/pullrequest.yml` to configure an action.

```yml
name: PR Helper

on: [pull_request, pull_request_review, issue_comment]

jobs:
  prhelper_job:
    runs-on: ubuntu-latest
    steps:
    - name: Run PR Helper
      id: runprhelper
      uses: Matticusau/pr-helper@v1.2.4
      with:
        repo-token: ${{ secrets.GITHUB_TOKEN }}
```

> Note: The `uses` syntax includes tag/branch specification. For the latest release see [tags](https://github.com/Matticusau/pr-helper/tags).

To restrict the branches that this workflow executes on use this syntax

```yml
name: PR Helper

on:
  pull_request:
    branches:
      - master
  pull_request_review:
    branches:
      - master
  issue_comment:
    branches:
      - master
jobs:
  ...
```

## Example inputs

The action can be customized using the additional inputs on the workflow yaml file. This will always be read from the default branch of the repository, rather than custom yaml config files which can be overridden as they are read in the branch where the workflow is triggered from.

```yml
with:
  repo-token: ${{ secrets.GITHUB_TOKEN }}
  enable-prmerge-automation: true
  enable-prcomment-automation: true
  enable-prlabel-automation: true
  enable-prreviewer-frontmatter: false
  enable-welcomemessage: true
  welcome-message: "Thanks for opening an issue! Make sure you've followed CONTRIBUTING.md."
  prmerge-requireallchecks: true
  prmerge-requirereviewcount: 1
  prmerge-method: 'merge'
  prlabel-default: 'pr-onhold'
  prlabel-ready: 'pr-ready'
  prlabel-onhold: 'pr-onhold'
  prlabel-reviewrequired: 'review-required'
  prlabel-automerge: 'auto-merge'
```

## Suggested Label Colors

Labels will be created during the assignment if they do not exist. The following are suggested labels and colors:

| Label | Description | Color |
| --- | --- | --- |
| pr-onhold | Pull Request is not yet ready to process automatically or review | #b60205 (red) |
| review-required | Pull Request or Issue requires review | #fbca04 (yellow) |
| qualifies-auto-merge | Pull Request qualifies for automatic merge | #0e8a16 (green) |
| pr-ready | Pull Request is ready to process or review | #0e8a16 (green) |

## Troubleshooting

If you are having issues running the action enable the debug logs as some additional logging has been built into the Action.

1. To enable runner diagnostic logging, set the following secret in the repository that contains the workflow: `ACTIONS_RUNNER_DEBUG` to `true`.
1. To download runner diagnostic logs, download the log archive of the workflow run. The runner diagnostic logs are contained in the `runner-diagnostic-logs` folder. For more information on downloading logs, see [Downloading logs](https://help.github.com/en/actions/configuring-and-managing-workflows/managing-a-workflow-run#downloading-logs).

[Enable debug logging](https://help.github.com/en/actions/configuring-and-managing-workflows/managing-a-workflow-run#enabling-debug-logging)

## Known issues

### PRs from Forked private repo [#24](https://github.com/Matticusau/pr-helper/issues/24)

From [https://docs.github.com/en/actions/reference/events-that-trigger-workflows#pull-request-events-for-forked-repositories](https://docs.github.com/en/actions/reference/events-that-trigger-workflows#pull-request-events-for-forked-repositories)

> Note: Workflows do not run on private base repositories when you open a pull request from a forked repository.

To work around this use the schedule event. This action supports the schedule event for Label, Review, and Merge features. Currently the Welcome message functionality is only supported for pull_request events.

Example YAML for running the action every day at 1am.

```yml
name: PR Merge on Schedule

on:
  schedule:
    - cron: '* 1 * * *'

jobs:
  prhelper_schedule:
    runs-on: ubuntu-latest
    steps:
    - name: Run PR Helper on Schedule
      id: runprhelperonschedule
      uses: Matticusau/pr-helper@v1.2.4
      with:
        repo-token: ${{ secrets.GHACTION_PAT }}
        enable-prmerge-automation: true
        enable-prcomment-automation: false
        enable-prlabel-automation: true
        enable-prreviewer-frontmatter: true
        enable-welcomemessage: false
        prmerge-requireallchecks: true
        prmerge-requirereviewcount: 1
        prmerge-method: 'merge'
        prmerge-deletebranch: 'true'
        prmerge-deletebranch-config: ''
        prmerge-pathcheck: true
        prmerge-allowpaths: '{"any":["articles/**"]}'
        prreviewer-authorkey: 'author'
        prreviewer-githubuserfromauthorfile: true
        prreviewer-authorfilepath: '_data/authors.yaml'
        prlabel-default: 'pr-onhold'
        prlabel-ready: 'pr-ready'
        prlabel-onhold: 'pr-onhold'
        prlabel-reviewrequired: 'review-required'
        prlabel-automerge: 'auto-merge'
```

### Multiple files and CODEOWNERS will block auto merge [#27](https://github.com/Matticusau/pr-helper/issues/27)

When using the CODEOWNERS feature and multiple files with different CODEOWNERS are modified. Even though you meet the minimum number of reviewers if there is still pending reviews from CODEOWNERS for files with no review yet, then this will block the merge and may throw an error on the action workflow.


RELEASES.md
# Releases

## v1.3.2

Releasing the following functionality:

- Includes some extra logging to troubleshoot issue in #59

Bug fixes:

- None

Breaking changes:

- None

## v1.3.1

Releasing the following functionality:

- Improved error handling for adding reviewers to prevent workflow failure trigger [Issue #55]

Bug fixes:

- None

Breaking changes:

- None

## v1.3.0

Releasing the following functionality:

- Added support for workflow_run event trigger

Bug fixes:

- None

Breaking changes:

- None

## v1.2.6

Releasing the following functionality:

- Additional info for troubleshooting

Bug fixes:

- None

Breaking changes:

- None

## v1.2.5

Releasing the following functionality:

- None

Bug fixes:

- Fixing the dependabot security alerts

Breaking changes:

- None

## v1.2.4

Releasing the following functionality:

- Bypassing reviewer check on files owned by PR author

Bug fixes:

- None

Breaking changes:

- None


## v1.2.3

Releasing the following functionality:

- None

Bug fixes:

- Improving the error handling of Not Found for #37
- Added logic to exclude frontmatter check on file rename #35

Breaking changes:

- None

## v1.2.2

Releasing the following functionality:

- None

Bug fixes:

- Frontmatter reviewer does not filter/skip attempting to assign the same user as PR owner which results in error #32

Breaking changes:

- None

## v1.2.1

Releasing the following functionality:

- None

Bug fixes:

- Error when FrontMatter reviewer is enabled but file has no frontmatter #30

Breaking changes:

- None

## v1.2.0

Releasing the following functionality:

- Auto merge not requiring qualify for auto merge label #28
- Action not running on PRs from Forked repos (excludes welcome message) #24
- Support for key value lookup file on front matter reviewers #23
- Auto merge filters (initial GLOB support) #15
- Documentation improvements
- Added docs on FrontMatter support

Bug fixes:

- Bump lodash from 4.17.15 to 4.17.19 #25 
- Action not running on PRs from Forked repos (excludes welcome message) #24

Breaking changes:

- Default `prlabel-automerge` label changed from `auto-merge` to `qualifies-auto-merge`. Use workflow config file to continue with your preferred label.

## v1.1.0

Releasing the following functionality:

- Documentation improvements

Breaking Changes:

- The configuration parameter `enable-prmerge-allowpaths` was renamed to `prmerge-allowpaths`

## v1.0.0

Releasing the following functionality:

- Delete branch on merge #14
- Docs on Auto merge filters #15
- Auto Merge Filter feature #18
- Assign reviewers from YAML front matter (e.g. jekyll) #19

Bug fixes:

- Auto merge should not proceed if approved reviews greater than minimum required but requested changes exist #16
- Multiple comment automation welcome messages #17

## v0.1

Initial release providing the following functionality:

- Welcome message on new PR
- Automatic Label assignment and removal
  - Based on Review criteria (required/provided)
  - If PRs qualify for automatic merge (___in preview___)
- Welcome message on new PR
- PR Comment automation
  - Welcome message with instructions on new PRs
  - Assign a label indicating if the PR is ready or on hold based on key words
- Pull Request Merge
  - Automatically merge when criteria is met
  - Respect requested reviews (_i.e. CODEOWNERS_)
  - Respect a minimal review count (_for now configured by input param, repo setting on roadmap_)



src/classes/prhelper.ts
//
// Author:  Matt Lavery
// Date:    2020-06-19
// Purpose: Helpers for working with pull requests
//
// When         Who         What
// ------------------------------------------------------------------------------------------
// 2020-06-22   Mlavery     Added check for Requested Changes in review #16
// 2020-06-24   MLavery     Improved logic by moving core and github to properties
// 2020-07-27   MLavery     Extended isMergeReadyByLabel() to check for qualifying label before merging [issue #28]
// 2021-02-08   Mlavery     Added check for isAllChangedFilesOwnedByPRAuthor() in isMergeReadyByReview() [issue #36]
//
import { CoreModule, GitHubModule,Context, PullRequestPayload } from '../types';
import { IssueLabels, PRFileHelper } from './index';
import { PullsGetResponseData } from '@octokit/types/dist-types'

interface DeleteBranchConfig {
    deny?: string[];
    allow?: string[];
}

export class PRHelper {
    
    // properties
    private core: CoreModule;
    private github: GitHubModule;
    mergemethod?: 'merge' | 'squash' | 'rebase'

    constructor(core: CoreModule, github: GitHubModule) {
        this.core = core;
        this.github = github;
        this.mergemethod = "merge";
    }

    setMergeMethod(method: string): void {
        if (method === 'squash') {
            this.mergemethod = 'squash';
        } else if (method === 'rebase') {
            this.mergemethod = 'rebase';
        } else {
            this.mergemethod = 'merge';
        }
    }

    getPrNumber(): number | undefined {
        var prNumber : number | undefined = undefined;
        // event will determine how we get this from the payload
        switch (this.github.context.eventName) {
            case 'issue_comment':
                const issue = this.github.context.payload.issue;
                if (issue) {
                    prNumber = issue.number;
                }
                break;
            default:
                const pullRequest = this.github.context.payload.pull_request;
                if (pullRequest) {
                    prNumber = pullRequest.number;
                } 
        }
        return prNumber;
    }

    getCommentNumber(): number | undefined {
        var commentNumber : number | undefined = undefined;
        // event will determine how we get this from the payload
        switch (this.github.context.eventName) {
            case 'issue_comment':
                const comment = this.github.context.payload.comment;
                if (comment) {
                    commentNumber = comment.id;
                }
                break;
            default:
                commentNumber = undefined;
        }
        return commentNumber;
    }

    isMergeReadyByState(pullRequest: PullsGetResponseData) : boolean {
        try {
            this.core.debug('>> isMergeReadyByState()');
            this.core.info('PR State: ' + pullRequest.state);
            this.core.info('PR merged: ' + pullRequest.merged);
            this.core.info('PR mergeable: ' + pullRequest.mergeable);
            this.core.info('PR mergeable_state: ' + pullRequest.mergeable_state);

            if (pullRequest.state !== 'closed') {
                if (pullRequest.merged === false && pullRequest.mergeable === true) {
                    // is ready to merge
                    if (pullRequest.mergeable_state === 'clean' || pullRequest.mergeable_state === 'unstable') {
                        return true;
                    }

                    // if blocked check for pending reviewers (this doesn't factor in that the review is approved or not)
                    if (pullRequest.mergeable_state === 'blocked' && (pullRequest.requested_reviewers.length === 0 && pullRequest.requested_teams.length === 0)) {
                        this.core.info(`PR #${pullRequest.number} is blocked but has no outstanding reviews`);
                        return true;
                    }

                    // allow blocked state - if branch protection is enabled the blocked state will be returned by the GitHub Actions user
                    if (pullRequest.mergeable_state === 'blocked') {
                        return true;
                    }
                }
            }
            return false;

        } catch (error) {
            this.core.info('isMergeReadyByState: Error caught and thrown.');
            this.core.setFailed(error.message);
            throw error;
        }
    }

    async isMergeReadyByLabel(pullRequest: PullsGetResponseData) : Promise<boolean> {
        try {
            this.core.debug('>> isMergeReadyByLabel()');

            const myToken = this.core.getInput('repo-token');
            const octokit = this.github.getOctokit(myToken);

            // check the labels
            const { data: issueLabelsData } = await octokit.issues.listLabelsOnIssue({
                ...this.github.context.repo,
                issue_number: pullRequest.number,
            });
            var issueLabels = new IssueLabels(issueLabelsData);
            const qualifiesForAutoMerge = (issueLabels.hasLabelFromList([this.core.getInput('prlabel-automerge')]));
            const readyToMergeLabel = (issueLabels.hasLabelFromList([this.core.getInput('prlabel-ready')]));
            const NotReadyToMergeLabel = (issueLabels.hasLabelFromList([this.core.getInput('prlabel-reviewrequired'), this.core.getInput('prlabel-onhold')]));
            
            this.core.info('qualifiesForAutoMerge: ' + qualifiesForAutoMerge);
            this.core.info('readyToMergeLabel:' + readyToMergeLabel);
            this.core.info('NotReadyToMergeLabel:' + NotReadyToMergeLabel);
            if (qualifiesForAutoMerge && readyToMergeLabel && !NotReadyToMergeLabel) {
                return true;
            }

            return false;

        } catch (error) {
            this.core.info('isMergeReadyByLabel: Error caught and thrown.');
            this.core.setFailed(error.message);
            throw error;
        }
    }

    async isMergeReadyByReview(pullRequest: PullsGetResponseData) : Promise<boolean> {
        try {
            this.core.debug('>> isMergeReadyByReview()');
            let result : boolean = false;

            // first we need to check if we should by pass if all files are owned by author
            const filehelper = new PRFileHelper(this.core, this.github);
            result = await filehelper.isAllChangedFilesOwnedByPRAuthor(pullRequest);
            this.core.info('isAllChangedFilesOwnedByPRAuthor result: ' + result);
            if ( !result ) {

                const myToken = this.core.getInput('repo-token');
                const octokit = this.github.getOctokit(myToken);
                const requiredReviewCount : Number = Number.parseInt(this.core.getInput('prmerge-requirereviewcount'));
                
                // check the labels
                const { data: reviewsData } = await octokit.pulls.listReviews({
                    ...this.github.context.repo,
                    pull_number: pullRequest.number,
                });

                let reviews = {
                    total: reviewsData.length,
                    approved: 0,
                    request_changes: 0
                };
                
                // No outstanding reviews (this doesn't check for approval vs comment so only valid if requiredReviewCount disabled)
                if (pullRequest.requested_reviewers.length === 0 && pullRequest.requested_teams.length === 0 && requiredReviewCount < 0) {
                    result = true;
                }
                
                // get the number of reviews
                for(var iReview = 0; iReview < reviewsData.length; iReview++){
                    if (reviewsData[iReview].state === 'APPROVED') {
                        reviews.approved++;
                    } else if (reviewsData[iReview].state === 'REQUEST_CHANGES') {
                        reviews.request_changes++;
                    } 
                }
                // check for reviews, and make sure no non-approved reviews
                // if (reviews.total > 0 && (reviews.total === reviews.approved) && ((requiredReviewCount >= 0 && reviews.approved >= requiredReviewCount) || requiredReviewCount < 0)) {
                if (reviews.total > 0 && (reviews.total === reviews.approved) && (requiredReviewCount >= 0 && reviews.approved >= requiredReviewCount)) {
                    this.core.info(`PR #${pullRequest.number} is mergable based on reviews`);
                    result = true;
                }

                // check for minimum number of required reviews (no requested changes)
                if (requiredReviewCount >= 0 && reviews.approved >= requiredReviewCount && reviews.request_changes === 0) {
                    this.core.info(`PR #${pullRequest.number} is mergable based on minimum required reviews`);
                    result = true;
                }
            }

            return result;

        } catch (error) {
            this.core.info('isMergeReadyByReview: Error caught and thrown.');
            this.core.setFailed(error.message);
            throw error;
        }
    }

    async isMergeReadyByChecks(pullRequest: PullsGetResponseData) : Promise<boolean> {
        try {
            this.core.debug('>> isMergeReadyByChecks()');
            
            const requireallchecks : boolean = (this.core.getInput('prmerge-requireallchecks') === 'true');

            // not configured
            if (requireallchecks !== true) {
                this.core.info('require checks is not enabled');
                return true;
            }
            
            // need all checks
            if (requireallchecks && await this.allChecksSucceeded(pullRequest)) {
                this.core.info('required checks have all succeeded');
                return true;
            }

            // failed test
            return false;

        } catch (error) {
            this.core.info('isMergeReadyByChecks: Error caught and thrown.');
            this.core.setFailed(error.message);
            throw error;
        }
    }

    async allChecksSucceeded(pullRequest: PullsGetResponseData) : Promise<boolean> {
        try {
            this.core.debug('>> allChecksPassed()');

            const myToken = this.core.getInput('repo-token');
            const octokit = this.github.getOctokit(myToken);

            // check the labels
            const { data: checksData } = await octokit.checks.listForRef({
                ...this.github.context.repo,
                ref: pullRequest.merge_commit_sha
            });

            let checks = {
                total: checksData.total_count,
                completed: 0,
                success: 0
            };

            this.core.info('checksData.check_runs.length: ' + checksData.check_runs.length);
            if (checksData && checksData.check_runs.length > 0) {
                checksData.check_runs.forEach(element => {
                    if (element.status === "completed") {
                        checks.completed++;
                    }
    
                    if (element.conclusion === "success") {
                        checks.success++;
                    }
                });
            }
            this.core.info('checks.total: ' + checks.total);
            this.core.info('checks.completed: ' + checks.completed);
            this.core.info('checks.success: ' + checks.success);

            return (checks.completed >= (checks.total - 1)) && (checks.success >= (checks.total - 1));

        } catch (error) {
            this.core.info('allChecksSucceeded: Error caught and thrown.');
            this.core.setFailed(error.message);
            throw error;
        }
    }

    private matchConfigFromActionInputYaml(json: string) : DeleteBranchConfig {
        try{
          // convert json to string array
          this.core.debug('json: ' + json);
          //let pattern : string[] = JSON.parse(json);
          let pattern : DeleteBranchConfig = JSON.parse(json);
          this.core.debug('json pattern: ' + JSON.stringify(pattern));
  
          return pattern;
        } catch (error) {
          this.core.info('matchConfigFromActionInputYaml: Error caught and thrown.');
          this.core.setFailed(error.message);
          throw error;
        }
    }

    async isBranchDeleteReady(pullRequest: PullsGetResponseData) : Promise<boolean> {
        try {
            this.core.debug('>> isBranchDeleteReady()');

            if (this.core.getInput('prmerge-deletebranch') === 'true') {
                if (pullRequest.head.repo.id === pullRequest.base.repo.id) {
                    // not the default branch
                    if (pullRequest.base.repo.default_branch !== pullRequest.head.ref) {
                        const deleteBranchConfig : DeleteBranchConfig = this.matchConfigFromActionInputYaml(this.core.getInput('prmerge-deletebranch-config'));
                        // check denies
                        if (deleteBranchConfig && deleteBranchConfig.deny && deleteBranchConfig.deny.length > 0) {
                            for(let iBranch = 0; iBranch < deleteBranchConfig.deny.length; iBranch++) {
                                // this.core.info(deleteBranchConfig.deny[iBranch]);
                                if (pullRequest.head.ref === deleteBranchConfig.deny[iBranch]) {
                                    // we found a match so fail
                                    return false;
                                }
                            }
                        }
                        // check allows
                        if (deleteBranchConfig && deleteBranchConfig.allow && deleteBranchConfig.allow.length > 0) {
                            for(let iBranch = 0; iBranch < deleteBranchConfig.allow.length; iBranch++) {
                                this.core.info(deleteBranchConfig.allow[iBranch]);
                                if (pullRequest.head.ref !== deleteBranchConfig.allow[iBranch]) {
                                    // we found a non-match so fail
                                    return false;
                                }
                            }
                        }

                        // check passed
                        return true;
                    } else {
                        this.core.info('Base and Head are same branch, delete branch skipped.')
                    }
                } else {
                    this.core.info('Base and Head not on same repo, delete branch skipped.')
                }
            }
            
            // failed test
            return false;

        } catch (error) {
            this.core.info('isBranchDeleteReady: Error caught and thrown.');
            this.core.setFailed(error.message);
            throw error;
        }
    }

    // async getChangedFiles(pullRequest: PullsGetResponseData): Promise<string[]> {
    //     this.core.debug('>> getChangedFiles()');

    //     const myToken = this.core.getInput('repo-token');
    //     const octokit = this.github.getOctokit(myToken);

    //     const listFilesResponse = await octokit.pulls.listFiles({
    //       owner: this.github.context.repo.owner,
    //       repo: this.github.context.repo.repo,
    //       pull_number: pullRequest.number
    //     });
      
    //     const changedFiles = listFilesResponse.data.map(f => f.filename);
      
    //     this.core.info('found changed files:');
    //     for (const file of changedFiles) {
    //       this.core.info('  ' + file);
    //     }
      
    //     return changedFiles;
    // }
}
